Panic in NewMessageFieldNode

Parsing this example, causes a panic in ast.NewMessageFieldNode:
```
syntax = "proto3";

package buf.lsp.test.v1;

import "buf/validate/validate.proto";

message TestMessage {
  TestNestedEnum test_nested_enum = 3 [(buf.validate.field) = {
    enum: {
      defined_only: true,
      not_in: [1,]
    }
  }];
}

enum TestEnum {
  TEST_ENUM_UNSPECIFIED = 0;
  TEST_ENUM_FOO = 1;
  TEST_ENUM_BAR = 1;
}
```

The stack trace is:
```
panic: val is nil

goroutine 26 [running]:
github.com/bufbuild/protocompile/ast.NewMessageFieldNode(0x1400010ee60, 0x140001274d0, {0x0, 0x0})
	/Users/alfred/go/pkg/mod/github.com/bufbuild/protocompile@v0.6.0/ast/values.go:541 +0x394
github.com/bufbuild/protocompile/parser.(*protoParserImpl).Parse(0x14000375000, {0x103559cd0, 0x14000051bc0})
	/Users/alfred/go/pkg/mod/github.com/bufbuild/protocompile@v0.6.0/parser/proto.y.go:1506 +0x56d4
github.com/bufbuild/protocompile/parser.protoParse({0x103559cd0, 0x14000051bc0})
	/Users/alfred/go/pkg/mod/github.com/bufbuild/protocompile@v0.6.0/parser/proto.y.go:966 +0x54
github.com/bufbuild/protocompile/parser.Parse({0x0, 0x0}, {0x103556b80, 0x1400041f580}, 0x140003fa6c0)
	/Users/alfred/go/pkg/mod/github.com/bufbuild/protocompile@v0.6.0/parser/parser.go:98 +0xd0
...
```

Ideally a ignorable error would be produced instead.

--- a/parser/parser_test.go
+++ b/parser/parser_test.go
@@ -67,6 +67,7 @@ func TestJunkParse(t *testing.T) {
 	for name, input := range inputs {
 		name, input := name, input
 		t.Run(name, func(t *testing.T) {
+			t.Parallel()
 			errHandler := reporter.NewHandler(reporter.NewReporter(
 				// returning nil means this will keep trying to parse after any error
 				func(err reporter.ErrorWithPos) error { return nil },

--- a/parser/parser_test.go
+++ b/parser/parser_test.go
@@ -51,18 +51,33 @@ func TestEmptyParse(t *testing.T) {
 
 func TestJunkParse(t *testing.T) {
 	t.Parallel()
-	errHandler := reporter.NewHandler(nil)
 	// inputs that have been found in the past to cause panics by oss-fuzz
 	inputs := map[string]string{
 		"case-34232": `'';`,
 		"case-34238": `.`,
+		"issue-196-a": `syntax = "proto3";
+		                message TestMessage {
+		                  option (ext) = { bad_array: [1,] }
+		                }`,
+		"issue-196-b": `syntax = "proto3";
+		                message TestMessage {
+		                  option (ext) = { bad_array [ , ] }
+		                }`,
 	}
 	for name, input := range inputs {
-		protoName := fmt.Sprintf("%s.proto", name)
-		_, err := Parse(protoName, strings.NewReader(input), errHandler)
-		// we expect this to error... but we don't want it to panic
-		require.Error(t, err, "junk input should have returned error")
-		t.Logf("error from parse: %v", err)
+		name, input := name, input
+		t.Run(name, func(t *testing.T) {
+			errHandler := reporter.NewHandler(reporter.NewReporter(
+				// returning nil means this will keep trying to parse after any error
+				func(err reporter.ErrorWithPos) error { return nil },
+				nil, // ignore warnings
+			))
+			protoName := fmt.Sprintf("%s.proto", name)
+			_, err := Parse(protoName, strings.NewReader(input), errHandler)
+			// we expect this to error... but we don't want it to panic
+			require.Error(t, err, "junk input should have returned error")
+			t.Logf("error from parse: %v", err)
+		})
 	}
 }
 

--- a/parser/proto.y
+++ b/parser/proto.y
@@ -492,22 +492,12 @@ listLiteral : '[' listElements ']' {
 	}
 
 listElements : listElement {
-		if $1 == nil {
-			$$ = nil
-		} else {
-			$$ = &valueSlices{vals: []ast.ValueNode{$1}}
-		}
+		$$ = &valueSlices{vals: []ast.ValueNode{$1}}
 	}
 	| listElements ',' listElement {
-		if $3 == nil {
-			$$ = $1
-		} else if $1 == nil {
-			$$ = &valueSlices{vals: []ast.ValueNode{$3}}
-		} else {
-			$1.vals = append($1.vals, $3)
-			$1.commas = append($1.commas, $2)
-			$$ = $1
-		}
+		$1.vals = append($1.vals, $3)
+		$1.commas = append($1.commas, $2)
+		$$ = $1
 	}
 
 listElement : scalarValue

--- a/parser/proto.y
+++ b/parser/proto.y
@@ -427,14 +427,14 @@ messageLiteralFieldEntry : messageLiteralField {
 	}
 
 messageLiteralField : messageLiteralFieldName ':' value {
-		if $1 != nil {
+		if $1 != nil && $2 != nil {
 			$$ = ast.NewMessageFieldNode($1, $2, $3)
 		} else {
 			$$ = nil
 		}
 	}
 	| messageLiteralFieldName messageValue {
-		if $1 != nil {
+		if $1 != nil && $2 != nil {
 			$$ = ast.NewMessageFieldNode($1, nil, $2)
 		} else {
 			$$ = nil
@@ -488,16 +488,26 @@ listLiteral : '[' listElements ']' {
 		$$ = ast.NewArrayLiteralNode($1, nil, nil, $2)
 	}
 	| '[' error ']' {
-		$$ = nil
+		$$ = ast.NewArrayLiteralNode($1, nil, nil, $3)
 	}
 
 listElements : listElement {
-		$$ = &valueSlices{vals: []ast.ValueNode{$1}}
+		if $1 == nil {
+			$$ = nil
+		} else {
+			$$ = &valueSlices{vals: []ast.ValueNode{$1}}
+		}
 	}
 	| listElements ',' listElement {
-		$1.vals = append($1.vals, $3)
-		$1.commas = append($1.commas, $2)
-		$$ = $1
+		if $3 == nil {
+			$$ = $1
+		} else if $1 == nil {
+			$$ = &valueSlices{vals: []ast.ValueNode{$3}}
+		} else {
+			$1.vals = append($1.vals, $3)
+			$1.commas = append($1.commas, $2)
+			$$ = $1
+		}
 	}
 
 listElement : scalarValue
@@ -514,7 +524,7 @@ listOfMessagesLiteral : '[' messageLiterals ']' {
 		$$ = ast.NewArrayLiteralNode($1, nil, nil, $2)
 	}
 	| '[' error ']' {
-		$$ = nil
+		$$ = ast.NewArrayLiteralNode($1, nil, nil, $3)
 	}
 
 messageLiterals : messageLiteral {

--- a/parser/proto.y.go
+++ b/parser/proto.y.go
@@ -1552,7 +1552,7 @@ protodefault:
 	case 68:
 		protoDollar = protoS[protopt-3 : protopt+1]
 		{
-			if protoDollar[1].ref != nil {
+			if protoDollar[1].ref != nil && protoDollar[2].b != nil {
 				protoVAL.msgLitFld = ast.NewMessageFieldNode(protoDollar[1].ref, protoDollar[2].b, protoDollar[3].v)
 			} else {
 				protoVAL.msgLitFld = nil
@@ -1561,7 +1561,7 @@ protodefault:
 	case 69:
 		protoDollar = protoS[protopt-2 : protopt+1]
 		{
-			if protoDollar[1].ref != nil {
+			if protoDollar[1].ref != nil && protoDollar[2].v != nil {
 				protoVAL.msgLitFld = ast.NewMessageFieldNode(protoDollar[1].ref, nil, protoDollar[2].v)
 			} else {
 				protoVAL.msgLitFld = nil
@@ -1624,19 +1624,29 @@ protodefault:
 	case 85:
 		protoDollar = protoS[protopt-3 : protopt+1]
 		{
-			protoVAL.v = nil
+			protoVAL.v = ast.NewArrayLiteralNode(protoDollar[1].b, nil, nil, protoDollar[3].b)
 		}
 	case 86:
 		protoDollar = protoS[protopt-1 : protopt+1]
 		{
-			protoVAL.sl = &valueSlices{vals: []ast.ValueNode{protoDollar[1].v}}
+			if protoDollar[1].v == nil {
+				protoVAL.sl = nil
+			} else {
+				protoVAL.sl = &valueSlices{vals: []ast.ValueNode{protoDollar[1].v}}
+			}
 		}
 	case 87:
 		protoDollar = protoS[protopt-3 : protopt+1]
 		{
-			protoDollar[1].sl.vals = append(protoDollar[1].sl.vals, protoDollar[3].v)
-			protoDollar[1].sl.commas = append(protoDollar[1].sl.commas, protoDollar[2].b)
-			protoVAL.sl = protoDollar[1].sl
+			if protoDollar[3].v == nil {
+				protoVAL.sl = protoDollar[1].sl
+			} else if protoDollar[1].sl == nil {
+				protoVAL.sl = &valueSlices{vals: []ast.ValueNode{protoDollar[3].v}}
+			} else {
+				protoDollar[1].sl.vals = append(protoDollar[1].sl.vals, protoDollar[3].v)
+				protoDollar[1].sl.commas = append(protoDollar[1].sl.commas, protoDollar[2].b)
+				protoVAL.sl = protoDollar[1].sl
+			}
 		}
 	case 90:
 		protoDollar = protoS[protopt-3 : protopt+1]
@@ -1655,7 +1665,7 @@ protodefault:
 	case 92:
 		protoDollar = protoS[protopt-3 : protopt+1]
 		{
-			protoVAL.v = nil
+			protoVAL.v = ast.NewArrayLiteralNode(protoDollar[1].b, nil, nil, protoDollar[3].b)
 		}
 	case 93:
 		protoDollar = protoS[protopt-1 : protopt+1]

--- a/parser/proto.y.go
+++ b/parser/proto.y.go
@@ -1629,24 +1629,14 @@ protodefault:
 	case 86:
 		protoDollar = protoS[protopt-1 : protopt+1]
 		{
-			if protoDollar[1].v == nil {
-				protoVAL.sl = nil
-			} else {
-				protoVAL.sl = &valueSlices{vals: []ast.ValueNode{protoDollar[1].v}}
-			}
+			protoVAL.sl = &valueSlices{vals: []ast.ValueNode{protoDollar[1].v}}
 		}
 	case 87:
 		protoDollar = protoS[protopt-3 : protopt+1]
 		{
-			if protoDollar[3].v == nil {
-				protoVAL.sl = protoDollar[1].sl
-			} else if protoDollar[1].sl == nil {
-				protoVAL.sl = &valueSlices{vals: []ast.ValueNode{protoDollar[3].v}}
-			} else {
-				protoDollar[1].sl.vals = append(protoDollar[1].sl.vals, protoDollar[3].v)
-				protoDollar[1].sl.commas = append(protoDollar[1].sl.commas, protoDollar[2].b)
-				protoVAL.sl = protoDollar[1].sl
-			}
+			protoDollar[1].sl.vals = append(protoDollar[1].sl.vals, protoDollar[3].v)
+			protoDollar[1].sl.commas = append(protoDollar[1].sl.commas, protoDollar[2].b)
+			protoVAL.sl = protoDollar[1].sl
 		}
 	case 90:
 		protoDollar = protoS[protopt-3 : protopt+1]

